### PR TITLE
Evaluation feature

### PR DESCRIPTION
**Pull request by LogicRunner**
Ability to define additional evaluations on the decision tree.

### CODE REVIEW

File: core/src/main/java/org/swblocks/decisiontree/Evaluator.java [hunk 1]
```diff
@@ -73,15 +72,31 @@ private Evaluator() {
      * @return List of highest weighted matching unique id's or {@code Collections.emptyList} if no match
      */
     public static List<UUID> evaluate(final List<String> searchInputs, final Instant time,
+                                      final Map<String, String> evaluationMap,
                                       final TreeNode rootNode) {
-        final List<EvaluationResult> results = evaluateAllResults(searchInputs, time, rootNode);
+        List<EvaluationResult> results = evaluateAllResults(searchInputs, time, rootNode);
```
Comment (jondey) [inline core/src/main/java/org/swblocks/decisiontree/Evaluator.java]: Formatting lost final keyword.

File: core/src/main/java/org/swblocks/decisiontree/Input.java [hunk 2]
```diff
@@ -38,18 +40,21 @@
...
     private Input(final String ruleSetName, final List<WeightedDriver> drivers, final List<String> searchValues,
-                  final Instant evaluationDate) {
+                  final Map<String, String> evaluations, final Instant evaluationDate) {
         this.ruleSetName = ruleSetName;
         this.evaluationDate = evaluationDate;
-        this.driverList = drivers;
-        this.driverMap = new TreeMap<>();
+        driverList = drivers;
+        driverMap = new TreeMap<>();
+        this.evaluationMap = evaluations;
```
Comment (jondey) [inline core/src/main/java/org/swblocks/decisiontree/Input.java:54]: evaluation assignment should be only code change.

File: core/src/main/java/org/swblocks/decisiontree/Input.java [hunk 3]
```diff
@@ -38,18 +40,21 @@
...
         int counter = 0;
         for (final WeightedDriver weightedDriver : drivers) {
-            this.driverMap.put(weightedDriver, searchValues.get(counter));
```
Comment (jondey) [inline core/src/main/java/org/swblocks/decisiontree/Input.java:52]: formatting again.

File: core/src/main/java/org/swblocks/decisiontree/Input.java [hunk 4]
```diff
@@ -106,7 +127,7 @@ static Input create(final String ruleSetName, final List<WeightedDriver> driverN
      * @return name of the ruleset
      */
     public String getRuleSetName() {
-        return this.ruleSetName;
+        return ruleSetName;
```
Comment (jondey) [inline core/src/main/java/org/swblocks/decisiontree/Input.java]: formatting...

File: core/src/main/java/org/swblocks/decisiontree/Input.java [hunk 5]
```diff
@@ -119,8 +140,8 @@ public String getRuleSetName() {
     public boolean putValueForDriverName(final String driverName, final String value) {
         // Check for valid key before accepting
         final WeightedDriver key = getWeightedDriverForDriverName(driverName);
-        if (key != null && this.driverMap.containsKey(key)) {
-            this.driverMap.put(key, value);
+        if (key != null && driverMap.containsKey(key)) {
```
Comment (jondey) [inline core/src/main/java/org/swblocks/decisiontree/Input.java]: formatting ...

File: core/src/main/java/org/swblocks/decisiontree/Input.java [hunk 6]
```diff
@@ -129,15 +150,15 @@ public boolean putValueForDriverName(final String driverName, final String value
 
     private WeightedDriver getWeightedDriverForDriverName(final String driverName) {
         // Check for valid key before accepting
-        final int driverPosition = this.driverList.indexOf(new WeightedDriver(driverName, 0));
+        final int driverPosition = driverList.indexOf(new WeightedDriver(driverName, 0));
```
Comment (jondey) [inline core/src/main/java/org/swblocks/decisiontree/Input.java]: should be no change to this method.

File: core/src/main/java/org/swblocks/decisiontree/Input.java [hunk 7]
```diff
@@ -129,15 +150,15 @@ public boolean putValueForDriverName(final String driverName, final String value
...
         if (driverPosition >= 0) {
-            return this.driverList.get(driverPosition);
+            return driverList.get(driverPosition);
```
Comment (jondey) [inline core/src/main/java/org/swblocks/decisiontree/Input.java]: formatting again..

File: core/src/main/java/org/swblocks/decisiontree/domain/DecisionTreeRule.java [hunk 8]
```diff
@@ -57,28 +59,46 @@
...
+
     /**
      * Returns true if the rule is active at the point in time passed in.
      *
      * @param time to check is rule is valid.
      * @return true is rule is active at time.
      */
     public boolean isActiveAt(final Instant time) {
-        return time.isAfter(this.start) && time.isBefore(this.end);
+        return time.isAfter(start) && time.isBefore(end);
```
Comment (jondey) [inline core/src/main/java/org/swblocks/decisiontree/domain/DecisionTreeRule.java]: formatting has removed this keyword from instance variables.

File: core/src/main/java/org/swblocks/decisiontree/domain/builders/RuleBuilder.java [hunk 9]
```diff
@@ -148,7 +154,7 @@ public void setAlternativeId(final Long alternativeId) {
      * @param uuid Unique Id of the rule
      */
     public void setCode(final UUID uuid) {
-        this.code = uuid;
+        code = uuid;
```
Comment (jondey) [inline core/src/main/java/org/swblocks/decisiontree/domain/builders/RuleBuilder.java]: general formatting has changed..

File: core/src/main/java/org/swblocks/decisiontree/tree/TimeSlicedRootNode.java [hunk 10]
```diff
@@ -55,7 +55,7 @@
     List<Instant> getTimes() {
         final Set<Instant> times = new TreeSet<>();
 
-        this.ruleSet.getRules().forEach((ruleId, rule) -> {
+        ruleSet.getRules().forEach((ruleId, rule) -> {
```
Comment (jondey) [inline core/src/main/java/org/swblocks/decisiontree/tree/TimeSlicedRootNode.java]: general formatting...

**Review by jondey — DISMISSED** [review]

**Review by jondey — COMMENTED** [review]
All good apart from formatting changes - this and final keyword changes.

File: core/src/main/java/org/swblocks/decisiontree/Evaluator.java (hunk 1 above)
Comment (jondey) [inline core/src/main/java/org/swblocks/decisiontree/Evaluator.java]: final...

File: core/src/main/java/org/swblocks/decisiontree/Input.java [hunk 11]
```diff
@@ -38,18 +40,21 @@
...
         this.evaluationDate = evaluationDate;
-        this.driverList = drivers;
-        this.driverMap = new TreeMap<>();
+        driverList = drivers;
+        driverMap = new TreeMap<>();
+        this.evaluationMap = evaluations;
         int counter = 0;
         for (final WeightedDriver weightedDriver : drivers) {
-            this.driverMap.put(weightedDriver, searchValues.get(counter));
+            driverMap.put(weightedDriver, searchValues.get(counter));
```
Comment (jondey) [inline core/src/main/java/org/swblocks/decisiontree/Input.java:57]: this

File: core/src/main/java/org/swblocks/decisiontree/Input.java (hunk 4 above)
Comment (jondey) [inline core/src/main/java/org/swblocks/decisiontree/Input.java]: formatting

File: core/src/main/java/org/swblocks/decisiontree/Input.java (hunk 5 above)
Comment (jondey) [inline core/src/main/java/org/swblocks/decisiontree/Input.java]: formatting...

File: core/src/main/java/org/swblocks/decisiontree/Input.java (hunk 6 above)
Comment (jondey) [inline core/src/main/java/org/swblocks/decisiontree/Input.java]: formatting

File: core/src/main/java/org/swblocks/decisiontree/Input.java (hunk 7 above)
Comment (jondey) [inline core/src/main/java/org/swblocks/decisiontree/Input.java]: formatting

File: core/src/main/java/org/swblocks/decisiontree/Input.java [hunk 12]
```diff
@@ -129,15 +150,15 @@ public boolean putValueForDriverName(final String driverName, final String value
...
         if (driverPosition >= 0) {
-            return this.driverList.get(driverPosition);
+            return driverList.get(driverPosition);
         }
         return null;
     }
 
     public String getValueForDriverName(final String driverName) {
-        return this.driverMap.get(getWeightedDriverForDriverName(driverName));
+        return driverMap.get(getWeightedDriverForDriverName(driverName));
```
Comment (jondey) [inline core/src/main/java/org/swblocks/decisiontree/Input.java:161]: formatting

File: core/src/main/java/org/swblocks/decisiontree/Input.java [hunk 13]
```diff
@@ -156,7 +177,7 @@ public boolean putBlankValueForDriverName(final String driverName) {
      * @return Instant
      */
     public Instant getEvaluationDate() {
-        return this.evaluationDate;
+        return evaluationDate;
```
Comment (jondey) [inline core/src/main/java/org/swblocks/decisiontree/Input.java:180]: ditto

File: core/src/main/java/org/swblocks/decisiontree/Input.java [hunk 14]
```diff
@@ -165,15 +186,19 @@ public Instant getEvaluationDate() {
...
                 Collections::unmodifiableList));
     }
 
     @Override
     public String toString() {
-        return "Input{driverMap=" + this.driverMap +
-                ", ruleSetName='" + this.ruleSetName + '\'' +
-                ", evaluationDate=" + this.evaluationDate +
+        return "Input{driverMap=" + driverMap +
+                ", ruleSetName='" + ruleSetName + '\'' +
```
Comment (jondey) [inline core/src/main/java/org/swblocks/decisiontree/Input.java:196]: re-formatted...

File: core/src/main/java/org/swblocks/decisiontree/Input.java [hunk 15]
```diff
@@ -165,15 +186,19 @@ public Instant getEvaluationDate() {
...
+                ", evaluationDate=" + evaluationDate +
                 '}';
     }
+
+    public Map<String, String> getEvaluationMap() {
+        return evaluationMap;
```
Comment (jondey) [inline core/src/main/java/org/swblocks/decisiontree/Input.java:202]: this.evaluationMAp

File: core/src/main/java/org/swblocks/decisiontree/domain/DecisionTreeRule.java (hunk 8 above)
Comment (jondey) [inline core/src/main/java/org/swblocks/decisiontree/domain/DecisionTreeRule.java]: formatter...

File: core/src/main/java/org/swblocks/decisiontree/domain/DecisionTreeRule.java [hunk 16]
```diff
@@ -111,12 +145,12 @@ public UUID getRuleIdentifier() {
      */
     @Override
     public Map<String, String> getOutputs() {
-        return Collections.unmodifiableMap(this.outputs);
+        return Collections.unmodifiableMap(outputs);
```
Comment (jondey) [inline core/src/main/java/org/swblocks/decisiontree/domain/DecisionTreeRule.java:148]: formatter

File: core/src/main/java/org/swblocks/decisiontree/domain/DecisionTreeRule.java [hunk 17]
```diff
@@ -111,12 +145,12 @@ public UUID getRuleIdentifier() {
...
     }
 
     @Override
     public DateRange getRange() {
-        return new DateRange(this.start, this.end);
+        return new DateRange(start, end);
```
Comment (jondey) [inline core/src/main/java/org/swblocks/decisiontree/domain/DecisionTreeRule.java:153]: formatter

File: core/src/main/java/org/swblocks/decisiontree/Evaluator.java (hunk 1 above)
Comment (LogicRunner) [reply core/src/main/java/org/swblocks/decisiontree/Evaluator.java]: it is final in latest code

**Review by jondey — APPROVED** [review]

**Pull request merged**